# Single-cell transcriptomic and epigenetic analysis of the human maternal-fetal interface with unexplained spontaneous miscarriage 
single cell transcriptome analysis  for villi sample in USM
This repository contains the scripts to reproduce the results of the manuscript “Single-cell transcriptomic and epigenetic analysis of the human maternal-fetal interface with unexplained spontaneous miscarriage”

Abstract
========
Whilst known genetic causes are identified in human pregnancy loss during early gestation, mechanisms underlying unexplained spontaneous miscarriage (USM) in patients are under-explored. Here, we perform single-cell transcriptomic and epigenetic analysis during human placental villi development in healthy donors and USM patients. Through constructing the dual-modality trajectory of trophoblast differentiation from cytotrophoblasts (CTBs) to extravillous trophoblasts (EVTs), we reveal that a subpopulation of CTBs undergoes precocious differentiation toward terminal EVTs with markedly decreased proliferation, invasion and cell-cell communications in USM patients. This defect is also accompanied by impaired epithelial–mesenchymal transition (EMT) process during placentation. Single-cell epigenomic profiling of putative enhancers by H3K27ac CoBATCH assay unravels dismantled regulation of key transcription factors (TFs), such as TEAD1 and SMAD3, in abnormal EVTs differentiation. Further, we identify a USM risk variant located at the trophoblast-specific gene SERPINE1 locus, whose mutation diminishes EMT TF ZEB1/2 occupancy on this newly identified enhancer. Together, our study provides a deep insight into cellular and epigenetic basis of USM placental villus formation.
![实验设计](https://user-images.githubusercontent.com/54924112/196712452-ea9497cc-89e9-421c-9a39-809b6bf61226.png)


Content
=======
/CoBATCH/: analysis of cobatch H3K27ac data

/scRNA/: analysis of single-cell RNA expression data

/Two omics intergration/: intergrated analysis of RNA expression and cobatch H3K27ac data

Data
======
The sequencing data have been deposited at the National Genomics Data Center database with the accession number HRA001886. 
Those data can be obtained upon request for researchers who meet the criteria for access to confidential data.

Contact
=======
Jie Qiao (jie.qiao@263.net) 
Aibin He (ahe@pku.edu.cn)
Chan Tian (tianchan@bjmu.edu.cn)
